sql/src/main/java/io/crate/metadata/rule/ingest/IngestionService.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata.rule.ingest;
+
+import io.crate.metadata.Schemas;
+import io.crate.metadata.TableIdent;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+@Singleton
+public class IngestionService implements ClusterStateListener {
+
+    private final ClusterService clusterService;
+    private final Set<IngestRuleListener> ingestRuleListeners = Collections.synchronizedSet(new HashSet<>());
+    private final Schemas schemas;
+
+    @Inject
+    public IngestionService(Schemas schemas,
+                            ClusterService clusterService) {
+        this.schemas = schemas;
+        this.clusterService = clusterService;
+    }
+
+    public void addListener(IngestRuleListener listener) {
+        ingestRuleListeners.add(listener);
+        listener.onRulesChanged(getIngestionRules().get(listener.getSourceIdent()));
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        Map<String, Set<IngestRule>> ingestionRules = getIngestionRules();
```

Comment:
event.state().meta() must be passed to the getIngestRules() to read from the actual state